types/share_merging.go

```diff
@@ -212,71 +212,65 @@ func parseMsgShares(shares [][]byte) ([]Message, error) {
 	if len(shares) == 0 {
 		return nil, nil
 	}
-
-	// set the first nid and current share
-	nid := shares[0][:consts.NamespaceSize]
-	currentShare := shares[0][consts.NamespaceSize:]
-	// find and remove the msg len delimiter
-	currentShare, msgLen, err := ParseDelimiter(currentShare)
-	if err != nil {
-		return nil, err
+	// msgs returned
+	msgs := []Message{}
+	// current msg len
+	msgLen := 0
+	// current msg
+	currentMsg := Message{}
+	// the current share contains the start of a new message
+	newMessage := true
+	// the len in bytes of the current chunk of data that will eventually become
+	// a message. This is identical to len(currentMsg.Data) + consts.MsgShareSize
+	// but we cache it here for readability
+	dataLen := 0
+
+	saveLatestMessage := func() {
+		msgs = append(msgs, currentMsg)
+		dataLen = 0
+		newMessage = true
 	}
-
-	var msgs []Message
-	for cursor := uint64(0); cursor < uint64(len(shares)); {
-		var msg Message
-		currentShare, nid, cursor, msgLen, msg, err = nextMsg(
-			shares,
-			currentShare,
-			nid,
-			cursor,
-			msgLen,
-		)
-		if err != nil {
-			return nil, err
-		}
-		if msg.Data != nil {
-			msgs = append(msgs, msg)
+	// iterate through all the shares and parse out each msg
+	for i := 0; i < len(shares); i++ {
+		dataLen = len(currentMsg.Data) + consts.MsgShareSize
+		switch {
+		case newMessage:
+			nextMsgChunk, nextMsgLen, err := ParseDelimiter(shares[i][consts.NamespaceSize:])
+			if err != nil {
+				return nil, err
+			}
+			// the current share is namespaced padding so we ignore it
+			if nextMsgLen == 0 {
+				continue
+			}
```

Comment:
This seems like it would work, since the bytes are zero and therefore the length delimiter is zero. But this also seems kind of brittle, since it's not really the length delimiter that's zero but all the bytes, specifically. Is there some way of exposing this information better, rather than relying on implementation details of the delimiter encoding?